revise: sc handle users who apply for alp after money is donated (TS-2217)

maikelordaz opened the pull request.

elallamTaka commented on contracts/referrals/ReferralGateway.sol:
```diff
@@ -669,36 +701,82 @@ contract ReferralGateway is
+
+            emit OnPrepayment(_parent, _member, normalizedContribution, _finalFee, _discount);
+        } else {
+            nameToDAOData[daoName]
+                .prepaidMembers[_member]
+                .contributionBeforeFee += normalizedContribution;
+            nameToDAOData[daoName].prepaidMembers[_member].contributionAfterFee +=
+                normalizedContribution -
+                (normalizedContribution * SERVICE_FEE_RATIO) /
+                100;
+            nameToDAOData[daoName].prepaidMembers[_member].feeToOperator += _finalFee;
+            nameToDAOData[daoName].prepaidMembers[_member].discount += _discount;
+            nameToDAOData[daoName].prepaidMembers[_member].isDonated = _isDonated;
```
Whats the benefit of summing the contributions? The referral gateways mission is to store the first annual contribution. It does not handle the payments for the new year. i.e pay recurring function does not exist. Thus, in my opinion, we should simply rewrite the info, no need for an if else here. Unless you think otherwise.

elallamTaka commented on contracts/referrals/ReferralGateway.sol:
```diff
@@ -365,38 +370,54 @@ contract ReferralGateway is
+     * @param prepaidMember The address of the prepaid member
+     * @param couponAmount The amount of the coupon. In USDC six decimals
+     */
+    function modifyPrepaidMember(
+        uint256 proRatedContribution,
+        address prepaidMember,
+        uint256 couponAmount
+    )
```
Lets look at the 2 prorated cases mentioned here: https://docs.google.com/document/d/1PiDGjVrNeWzB6D_EScrQ9wSEppSljXpACmcsoCa-K_w/edit?usp=sharing

For case 1, will this called as `modifyPrepaidMember(37.6, member, 50)` or as `modifyPrepaidMember(50, member, 37.6)` ?

maikelordaz commented on contracts/referrals/ReferralGateway.sol:
```diff
@@ -77,6 +77,8 @@ contract ReferralGateway is
 
     string public daoName;
 
+    uint256 public donationsFromCoupons; // In USDC, six decimals
```
Commit a722b5352022dc7e13f35b7ce64a500d6e18c72c

maikelordaz commented on contracts/referrals/ReferralGateway.sol:
```diff
@@ -365,38 +370,54 @@ contract ReferralGateway is
+        returns (uint256 finalFee, uint256 discount)
+    {
+        uint256 proRatedCouponAmount;
+        uint256 donationFromProRatedCoupon;
```
Commit a722b5352022dc7e13f35b7ce64a500d6e18c72c

maikelordaz commented:
@elallamTaka I changed the behaviour so not an exact commit for each comment you made.
- a722b5352022dc7e13f35b7ce64a500d6e18c72c and dcf26e9c2a2a18cad2b7ccde95c9c66657d730b7 just some renames
- 4c16fe0f68887af280f478892f497988c731c4c3 timestamp as input
- 3f67268c9629511995e2bdbdaa0c8dc51ebd9057 just update internal calls to be able to compile
- d3f514dcd69f5665f572fdfb7c3f36438ac1a6c2 this one is the one I changed the behaviour
- 7276439bf4046fced6ec735e68d5829b0497a29a and this one I fixed an underflow with the coupons

elallamTaka commented on contracts/referrals/ReferralGateway.sol:
```diff
@@ -651,71 +701,125 @@ contract ReferralGateway is
+        uint256 _associationTimestamp
     ) internal view {
-        // The payer must be different than the zero address and cannot be already a member
+        if (!_isModifying) _checksIfNotModifying(_newMember, _contribution, _couponAmount);
```
Simplify the reading of the code by avoiding negation when possible. Since the book checks for modification: `(_isModifying)  _checksIfModifying(_newMember, _couponAmount, _associationTimestamp); ...`

maikelordaz replied:
Commit d1b24c83d7ff09852a8a73579497446eb8486c8e

elallamTaka commented on contracts/referrals/ReferralGateway.sol:
```diff
@@ -365,38 +370,42 @@ contract ReferralGateway is
+     * @param couponAmount The amount of the coupon. In USDC six decimals
+     * @param associationTimestamp The timestamp the first payment was made
+     */
+    function lateContribution(
```
To follow the standard of starting with a verb first (like payContribution) I recommend we rename this to `payLateContribution` or `payContributionAfterWaive` for better context

maikelordaz replied:
Commit d1b24c83d7ff09852a8a73579497446eb8486c8e

maikelordaz commented on contracts/referrals/ReferralGateway.sol:
```diff
@@ -651,71 +701,125 @@ contract ReferralGateway is
-            ReferralGateway__InvalidContribution()
-        );
-        if (_couponAmount > 0)
+        if (_couponAmount > 0) {
```
True, but it is harmless and good for EOAs in the future, although dont think we'll still be using the `ReferralGateway` at that moment

maikelordaz commented on contracts/referrals/ReferralGateway.sol:
```diff
@@ -365,38 +370,42 @@ contract ReferralGateway is
-        emit OnMemberKYCVerified(child);
+        // Set as a coupon redeemer if not already
+        if (couponAmount > 0 && !isMemberCouponRedeemer[prepaidMember])
+            _assignAsCouponRedeemer(prepaidMember, couponAmount);
```
Same as the other comment harmless and usefull for EOAs. I'll remove those if blocks if I need to release some space

maikelordaz commented on contracts/referrals/ReferralGateway.sol:
```diff
@@ -651,71 +701,125 @@ contract ReferralGateway is
+            // If we are not modifying, we create a new prepaid member
+            nameToDAOData[daoName].prepaidMembers[_member].member = _member;
+
+            emit OnPrepayment(_parent, _member, normalizedContribution, _finalFee, _discount);
```
Same comment as the previous one

maikelordaz commented on contracts/referrals/ReferralGateway.sol:
```diff
@@ -651,71 +701,125 @@ contract ReferralGateway is
+        nameToDAOData[daoName].prepaidMembers[_member].contributionAfterFee =
             normalizedContribution -
             (normalizedContribution * SERVICE_FEE_RATIO) /
             100;
```
Some comments in ac72efe89a63bc7eb9a2c495dcf3ac1f683c27d4 rename inputs in f21f9e324d369b6027ff575b5f52047106fe97df

maikelordaz commented on contracts/referrals/ReferralGateway.sol:
```diff
@@ -651,71 +701,125 @@ contract ReferralGateway is
+        require(
+            _contribution >= MINIMUM_CONTRIBUTION && _contribution <= MAXIMUM_CONTRIBUTION,
+            ReferralGateway__InvalidContribution()
+        );
```
Actually it is a wording issue, in this caase the contribution should be the plan commit c7f9bcfd8a7468aba7c485c505f8db85221f020d

elallamTaka commented on contracts/referrals/ReferralGateway.sol:
```diff
@@ -607,115 +649,229 @@ contract ReferralGateway is
+        uint256 _benefitStartsAt
+    ) internal pure returns (uint256) {
+        uint256 YEAR = 365 days;
+        if (_associationStartsAt >= _benefitStartsAt) return _amount;
```
My preference is for the call to fail if this is true, because it means the backend called the function with the wrong timestamps, which is a bug.

elallamTaka commented on contracts/referrals/ReferralGateway.sol:
```diff
@@ -607,115 +649,229 @@ contract ReferralGateway is
+            totalDonationsFromCoupons += _inputs.couponAmount - proratedContribution;
+        } else {
+            fixedCouponAmount = _inputs.couponAmount;
+        }
```
To confirm, the backend must pass in the coupon amount as the original plan for users who pay prorated correct? For example, if a user applies to the 25$ plan (without a coupon on the FE), then it should be called as `payContributionAfterWaive(25, member, 25, timestamp, timestamp) `?

This means that coupon amount must always be equal to the contribution if the backend is the caller.

elallamTaka replied:
The reason I want to confirm this is because if its not the case, then the function will try to transfer USDC from the members wallet, which does not work when the backend is the called.

maikelordaz commented on contracts/referrals/ReferralGateway.sol:
```diff
@@ -326,23 +351,30 @@ contract ReferralGateway is
+        //     _isModifying: false, // In this call we never modify, as this will be a new member
+        //     _associationTimestamp: 0, // Only used for payContributionAfterWaive
+        //     _benefitTimestamp: 0 // Only used for payContributionAfterWaive
+        // });
```
commit 8d6022ec5d827e7f371eb0872edfb0f0cc978eac

maikelordaz commented on contracts/referrals/ReferralGateway.sol:
```diff
@@ -607,115 +649,229 @@ contract ReferralGateway is
+            // If this is modifying, it means the member already exists, and can have parents
+            // We transfer the rewards if needed
+            if (isMemberKYCed[_inputs.member]) _transferRewardsFromChild(_inputs.member);
+            emit OnPrepaidMemberModified(normalizedContribution, finalFee_, discount_);
```
commit 0c889060d7abaecb6685d49fd70e494f5fee2630

maikelordaz commented on contracts/referrals/ReferralGateway.sol:
```diff
@@ -607,115 +649,229 @@ contract ReferralGateway is
+        if (_associationStartsAt >= _benefitStartsAt) return _amount;
+        uint256 elapsed = _benefitStartsAt - _associationStartsAt;
+        if (elapsed >= YEAR) return 0;
+        return (_amount * (YEAR - elapsed)) / YEAR;
```
In that case this function returns 6$ not 18$. The remaining 100 would be `YEAR - elapsed` not the `elapsed` the return should be `_amount - ((_amount * elapsed) / YEAR)` or `_amount * remaining / YEAR` where `remaining = YEAR - elapsed`

elallamTaka commented on test/unit/05-referral-gateway/14-ModifyingMemberTest.t.sol:
```diff
@@ -0,0 +1,314 @@
+
+        // The child pays the minimum contribution without a coupon, using the parent as a referrer, and donates the contribution
+        vm.prank(couponRedeemer);
+        referralGateway.payContributionOnBehalfOf(25e6, parent, child, 0, true);
```
I dont think its worth focusing on testing with the coupon amount as 0 because it will simply never be the case in the past months, and it will never the case for the referral gateway. The coupon amount must be equal to the contribution to mimic the backends behavior.

elallamTaka commented on test/unit/05-referral-gateway/14-ModifyingMemberTest.t.sol:
```diff
@@ -0,0 +1,314 @@
+        assert(isDonated);
+
+        uint256 newContribution = 50e6;
+        uint256 couponAmount = 100e6; // 100 USDC coupon
```
This does not mimic the backend behavior. Pay contribution on behalf and pay contribution after waive both always have the same contribution and coupon amounts as input. While these are valid inputs from the code perspective, and its valid to test them, we should still focus our tests on how the backend is meant to call them. 

While you can keep these tests, the following should also be tested (contribution == coupon amount): 
* User joins after waive with a 25$ plan && 100 days left for membership (with parent)
* User joins after waive with a 100$ plan && 200 days left for membership (without parent)
* User joins after waive with a 250$ plan and 350 days left for membership (without parent). 
* User should not be able to double waive.

elallamTaka reviewed (changes requested):
Hello Maikel, the logic seems sound! Well done! Whats left is to perfect the tests. The tests are organized but i dont think they focus on how the backend makes the calls. Also, there are more things that can be asserted per test, ex. the actual amount a parent gets, and members balance or coupon balance before and after.

maikelordaz commented on test/unit/05-referral-gateway/14-ModifyingMemberTest.t.sol:
```diff
@@ -0,0 +1,314 @@
+        assertEq(discount, 6_575_341);
+        assert(!isDonated);
+
+        assert(parentBalanceAfter > parentBalanceBefore);
```
commit f94017741e0f8be1283e9b86ea8f33b02f92136f and 84a671dbf37db764cb568f9ab6ee0a5ac471edf3

maikelordaz commented on test/unit/05-referral-gateway/14-ModifyingMemberTest.t.sol:
```diff
@@ -0,0 +1,314 @@
+
+        // The parent pays the full contribution without a coupon
+        vm.prank(couponRedeemer);
+        referralGateway.payContributionOnBehalfOf(MAX_CONTRIBUTION, address(0), parent, 0, false);
```
commit c980cbdc4a83856f60cfd7c56141a562044e5045

elallamTaka commented on test/unit/05-referral-gateway/14-ModifyingMemberTest.t.sol:
```diff
@@ -0,0 +1,565 @@
+
+        // The child pays the minimum contribution without a coupon, using the parent as a referrer, and donates the contribution
+        vm.prank(couponRedeemer);
+        referralGateway.payContributionOnBehalfOf(25e6, parent, child, 0, true);
```
When donating, the backend calls coupon amount == contribution. Please update it here.

elallamTaka commented on test/unit/05-referral-gateway/14-ModifyingMemberTest.t.sol:
```diff
@@ -0,0 +1,565 @@
+        assertEq(discount, 375e4);
+        assert(isDonated);
+
+        // The child now wants to upgrade to the 50 USDC plan, so the new contribution is prorated
```
comment must say 25 USDC plan

maikelordaz replied:
Commit 9de930d7baeb9d424473130363ab533fed4533fc

maikelordaz commented on test/unit/05-referral-gateway/14-ModifyingMemberTest.t.sol:
```diff
@@ -0,0 +1,565 @@
+            MAX_CONTRIBUTION,
+            false
+        );
+        referralGateway.payContributionOnBehalfOf(25e6, address(0), parentLess, 25e6, true);
```
Commit 7415902dbecfd797ee6b6ad263e4f4dcd079fa66
